Engine/FindAstPositionVisitor.cs

```diff
@@ -0,0 +1,372 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Management.Automation.Language;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer
+{
+    /// <summary>
+    /// Provides an efficient way to find the position in the AST corresponding to a given script position.
+    /// </summary>
+#if !(PSV3 || PSV4)
+    internal class FindAstPositionVisitor : AstVisitor2
+#else
+    internal class FindAstPositionVisitor : AstVisitor
+#endif
+    {
+        private IScriptPosition searchPosition;
+
+        /// <summary>
+        /// Contains the position in the AST corresponding to the provided <see cref="IScriptPosition"/> upon completion of the <see cref="Ast.Visit(AstVisitor)"/> method.
+        /// </summary>
+        public Ast AstPosition { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FindAstPositionVisitor"/> class with the postition to search for.
+        /// </summary>
+        /// <param name="position">The script position to search for.</param>
+        public FindAstPositionVisitor(IScriptPosition position)
+        {
+            this.searchPosition = position;
+        }
+
+        /// <summary>
+        /// Traverses the AST based on offests to find the leaf node which contains the provided <see cref="IScriptPosition"/>.
+        /// This method implements the entire functionality of this visitor. All <see cref="AstVisitor2"/> methods are overridden to simply invoke this one.
+        /// </summary>
+        /// <param name="ast">Current AST node to process.</param>
+        /// <returns>An <see cref="AstVisitAction"/> indicating whether to visit children of the current node.</returns>
+        private AstVisitAction Visit(Ast ast)
+        {
+            if (ast.Extent.StartOffset > searchPosition.Offset || ast.Extent.EndOffset <= searchPosition.Offset)
+            {
+                return AstVisitAction.SkipChildren;
+            }
+            else
+            {
+                AstPosition = ast;
+                return AstVisitAction.Continue;
+            }
```

Comment:
This can be slightly simplified:
```suggestion
            AstPosition = ast;
            return AstVisitAction.Continue;
```

Reply:
The intent of this whole thing is to fix the problem described in #769. Briefly, if you enable `UseConsistentWhitespace` rule with `CheckOperator = $true` and `AlignAssignmentStatement` rule with `CheckHashtable = $true`, you get a conflict. Fix the warning given by one of the rules and instead you get a warning from the other.

If a user enables both of these settings, as I did right off the bat (and as is done in the various `CodeFormatting*` presets), my assumption would be that the user intends there to be exactly one space around assignment operators, except in hashtables, where they should be aligned. If I could figure out how to, instead of making this a configurable setting at all, I would have instead put code in somewhere that automatically enables this behavior if and only if `AlignAssignmentStatement` rule is also enabled with `CheckHashtable = $true`.

I don't think this change really makes sense in any other configuration. If `CheckOperator` is not set, the code won't run anyway. If `CheckOperator` is set but `AlignAssignmentStatement` is not enabled, or `CheckHashTable` is not set, then by enabling `IgnoreAssignmentOperatorInsideHashTable` you are basically saying that a single space is required around assignment operators, except within hashtables where anything goes (does not need to have whitespace or be aligned).

If you think it would be best to keep it as a user-configurable setting as I have done, I think the default should always reflect the defaults of `CheckOperator` and `CheckHashTable`. If either of those defaults to `$false`, then so should `IgnoreAssignmentOperatorInsideHashTable`. I put the default value of `IgnoreAssignmentOperatorInsideHashTable` as false initially both because
- it definitely won't break anything that way
- and, [the rule documentation for AlignAssignmentStatement](https://github.com/PowerShell/PSScriptAnalyzer/blob/1.19.1/RuleDocumentation/AlignAssignmentStatement.md#checkhashtable-bool-default-value-is-false) shows that default of `CheckHashTable` is `$false`, which now looking at [the code](https://github.com/PowerShell/PSScriptAnalyzer/blob/8c5a6388cc121c8308217039e9d1f9bba4aa4ff2/Rules/AlignAssignmentStatement.cs#L31) , I see actually the default is true

I guess since both `CheckOperator` and `CheckHashtable` actually default to `true`, it would make sense to change `IgnoreAssignmentOperatorInsideHashTable` to default to `true` at some point. But I wouldn't remove the setting altogether, unless the behavior is made to depend on the settings of `AlignAssignmentStatement`.

I will go ahead and make the suggested simplification.

Reply:
Hmm, yes you are right about the choice of defaults, thanks for explaining.